### PR TITLE
Fix Undo/Redo Bugs

### PR DESCRIPTION
**Pull request by angkoonhwee**
- [x] Remains on filtered list after every command except add command #267 #240 #251 #259 
- [x] Undo and Redo will not mix up lists #259 #265 #265 
- [x] Detailed Messages of which command type was undone for which student #129

**Note:**
I don't quite get the implementation of `LessonEditTest#execute_filteredList_success`  but it is not passing. What happens now is that the expectedModel's filteredList is the unfilteredList whereas the actual model's filteredList is just the filtered student.
The thing is I am not sure where in the code was the model filtered.

### CODE REVIEW

**Comment by angkoonhwee** [comment]
@eeliana You might want to explain the `LessonEditCommandTest#execute_filteredList_success` to me so that I can change accordingly. Seems like the expectedModel is the one that is weird.

**Comment by eeliana** [comment]
> @eeliana You might want to explain the `LessonEditCommandTest#execute_filteredList_success` to me so that I can change accordingly. Seems like the expectedModel is the one that is weird.

You have to add `showPersonAtIndex(expectedModel, INDEX_FIRST_PERSON);` after expectedModel.setPerson... because now the list no longer goes back to unfiltered state after execution.

**Comment by lingshanng** [comment]
I found a bug. Given a list of 4 students
1. delete 4
2. undo
3. delete 4
4. redo

cannot redo 1. because only 3 students

**Comment by lingshanng** [comment]
Also, when you undo a command for a person not currently in the filtered list, the student's lessons appear at the side bar, even if it's not lesson command but a student command.

Eg.
1. edit Alex
2. Filter list without Alex
3. undo
-> Alex's lessons appear at the side

image

**Review by Chesterwongz — COMMENTED** [review]
I think only undoable commands show clear the undostack. Right now if i do this:
1. edit person
2. undo
3. `calendar`
4. redo
Result is no commands to redo, instead of redoing edit person

File: src/main/java/seedu/address/logic/commands/UndoCommand.java [hunk 1]
```diff
@@ -23,8 +25,23 @@ public CommandResult execute() throws CommandException {
...
             throw new CommandException(MESSAGE_FAILURE);
         }
-        undoRedoStack.popUndo().undo();
-        return new CommandResult(MESSAGE_SUCCESS);
+
+        UndoableCommand commandToUndo = undoRedoStack.popUndo();
+
+        Person studentModified = commandToUndo.undo();
+
+        if (studentModified != null && !model.getFilteredPersonList().contains(studentModified)) {
```
Comment (Xiaoyunnn) [inline src/main/java/seedu/address/logic/commands/UndoCommand.java:33]: maybe can create another method in model to make `!model.getFilteredPersonList().contains(studentModified)` obey SLAP

**Review by Xiaoyunnn — COMMENTED** [review]
I think it looks ok now, just need to specify clearly when can undo/redo command in ug, cus its a bit confusing now

**Comment by Xiaoyunnn** [comment]
small cosmetic issue:
1. delete 3 (charlotte)
2. undo (selects Charlotte's card)
3. redo (delete Charlotte, selects the prev student Bernice's card, but lesson panel shows generic student lesson list)
image

can make it not select prev person's card?

**Comment by Chesterwongz** [comment]
> cosmetic issue/bug(?):
> 
> 1. delete 3 (charlotte)
> 2. undo (selects Charlotte's card)
> 3. redo (delete Charlotte, selects the prev student Bernice's card, but lesson panel shows generic student lesson list, and when u click bernice’s card, it doesn't show bernice’s lessons)
>    image
> 
> can make it not select prev person's card?

Fixed in #279

**Review by Chesterwongz — APPROVED** [review]
LGTM

**Review by Xiaoyunnn — APPROVED** [review]
LGTM

**Comment by lingshanng** [comment]
Can I just check are these behaviours intended?

Add person & edit person -> doesn't select person and view lessons
Filter list and edit person -> goes back to full list

**Comment by angkoonhwee** [comment]
@lingshanng For point 2, I think that was the initial implementation of AB3 I don't think anyone changed it, but it would make more sense to stay on the list? I can change that while I am at it. As for point 1, I am not sure who implemented the command result thingy so it might have be intended. I personally think that selecting the add/edited student makes more sense and if it is agreeable I can also change that while I am at it.

**Comment by eeliana** [comment]
> @lingshanng For point 2, I think that was the initial implementation of AB3 I don't think anyone changed it, but it would make more sense to stay on the list? I can change that while I am at it. As for point 1, I am not sure who implemented the command result thingy so it might have been intended. I personally think that selecting the add/edited student makes more sense and if it is agreeable I can also change that while I am at it.

Just have to add toAdd and personAfterEdit to the CommandResult returned for AddCommand and EditCommand. I think we discussed in tele before.

**Review by eeliana — CHANGES_REQUESTED** [review]
Can add changes suggested, thanks for the hardwork!

**Review by lingshanng — APPROVED** [review]

**Pull request merged**